src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -560,6 +563,150 @@ private async ValueTask SendRstStreamAsync(int streamId, Http2ProtocolErrorCode
                 (buffer.Slice(0, FrameHeader.MaxLength), buffer.Slice(FrameHeader.MaxLength)) :
                 (buffer, Memory<byte>.Empty);
 
+        private void WriteHeader(string name, string value)
+        {
+            // TODO: ISSUE 31307: Use static table for known headers
+
+            int bytesWritten;
+            while (!HPackEncoder.EncodeHeader(name, value, _headerBuffer.AvailableSpan, out bytesWritten))
+            {
+                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableSpan.Length + 1);
+            }
+
+            _headerBuffer.Commit(bytesWritten);
+        }
+
+        private void WriteHeaders(HttpHeaders headers)
+        {
+            foreach (KeyValuePair<HeaderDescriptor, string[]> header in headers.GetHeaderDescriptorsAndValues())
+            {
+                if (header.Key.KnownHeader == KnownHeaders.Host)
+                {
+                    continue;
+                }
+
+                Debug.Assert(header.Value.Length > 0, "No values for header??");
+                for (int i = 0; i < header.Value.Length; i++)
+                {
+                    WriteHeader(header.Key.Name, header.Value[i]);
+                }
+            }
+        }
+
+        private void WriteHeaders(HttpRequestMessage request)
+        {
+            Debug.Assert(_headerBuffer.ActiveMemory.Length == 0);
+
+            // HTTP2 does not support Transfer-Encoding: chunked or Connection: headers.
+            // Simply clear these out so they won't be sent.
+            request.Headers.TransferEncodingChunked = false;
+            request.Headers.Connection.Clear();
```

Comment:
It's kind of painful to skip Transfer-Encoding: chunked because we can't just skip the whole Transfer-Encoding header as there could be other values for Transfer-Encoding. And given that we modify TransferEncodingChunked in other places, I'm inclined to leave as-is.

I will change the Connection handling to just skip it.

Reply:
That looks good, thanks.